cli/util/uploader.go

```diff
@@ -0,0 +1,432 @@
+	prm.SetFilters(filters)
+
+	objectIDs, err := neofs.ObjectSearch(ctx.Context, &p, account.PrivateKey(), containerID.String(), prm)
+	if err != nil {
+		return fmt.Errorf("search of index files failed: %w", err)
+	}
+
+	existingIndexCount := uint(len(objectIDs))
+	expectedIndexCount := currentHeight / indexFileSize
+
+	if existingIndexCount >= expectedIndexCount {
+		fmt.Fprintf(ctx.App.Writer, "Index files are up to date. Existing: %d, expected: %d\n", existingIndexCount, expectedIndexCount)
+		return nil
+	}
+
+	var (
+		processedCounter atomic.Int32
+		errCh            = make(chan error, 1)
+		buffer           = make([]byte, indexFileSize*oidSize)
+		oidCh            = make(chan oid.ID, indexFileSize)
+	)
+
+	for range maxParallelSearches {
+		go func() {
+			for currOid := range oidCh {
+				obj, err := p.ObjectHead(context.Background(), containerID, currOid, signer, client.PrmObjectHead{})
+				if err != nil {
+					select {
+					case errCh <- fmt.Errorf("failed to get header for OID %s: %w", currOid.String(), err):
+					default:
+					}
+					return
+				}
+
+				blockIndex, err := getBlockIndex(*obj, blockAttributeKey)
+				if err != nil {
+					select {
+					case errCh <- fmt.Errorf("failed to process header for OID %s: %w", currOid.String(), err):
+					default:
+					}
+					return
+				}
+
+				index := uint(blockIndex) / indexFileSize
+				offset := (uint(blockIndex) - index*indexFileSize) * oidSize
+				res := make([]byte, oidSize)
+				currOid.Encode(res)
+				copy(buffer[offset:], res)
+				processedCounter.Add(1)
+			}
+		}()
+	}
+
+	for i := existingIndexCount; i < expectedIndexCount; i++ {
+		startIndex := i * indexFileSize
+		endIndex := startIndex + indexFileSize
+
+		for j := int(startIndex); j < int(endIndex); j += searchBatchSize {
+			remaining := int(endIndex) - j
+			end := j + min(searchBatchSize, remaining)
+			prm := client.PrmObjectSearch{}
+			filters := object.NewSearchFilters()
+			filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", j), object.MatchNumGE)
+			filters.AddFilter(blockAttributeKey, fmt.Sprintf("%d", end), object.MatchNumLE)
+			prm.SetFilters(filters)
+
+			objectIDs, err = neofs.ObjectSearch(ctx.Context, &p, account.PrivateKey(), containerID.String(), prm)
+			if err != nil {
+				return fmt.Errorf("no OIDs found for index files: %w", err)
+			}
+
+			for _, currOid := range objectIDs {
+				oidCh <- currOid
+			}
+		}
+
+		for processedCounter.Load() < int32(indexFileSize) {
+			select {
+			case err := <-errCh:
+				return err
+			default:
+				time.Sleep(100 * time.Millisecond)
+			}
+		}
+
+		processedCounter.Store(0)
+
+		select {
+		case err := <-errCh:
+			return err
+		default:
+		}
+		attrs := []object.Attribute{
+			*object.NewAttribute(attributeKey, strconv.Itoa(int(i))),
+			*object.NewAttribute("size", strconv.Itoa(int(indexFileSize))),
+			*object.NewAttribute("timestamp", strconv.FormatInt(time.Now().Unix(), 10)),
+			*object.NewAttribute("block-attribute", blockAttributeKey),
```

Comment:
@AnnaShaleva what do you think about adding two more attributes to index files? `timestamp` seems to me now useless -NeoFS will put it by itself.  But block-attribute could be helpful.

Reply:
Timestamp should be a block timestamp, not the upload timestamp. Previous version of this PR had the correct code.